packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,266 @@
+import type { Libp2p } from "@libp2p/interface-libp2p";
+import type { Peer } from "@libp2p/interface-peer-store";
+import { IncomingStreamData } from "@libp2p/interface-registrar";
+import type {
+  ActiveSubscriptions,
+  Callback,
+  IDecodedMessage,
+  IDecoder,
+  IFilter,
+  IProtoMessage,
+  ProtocolCreateOptions,
+  ProtocolOptions,
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import debug from "debug";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { BaseProtocol } from "../../base_protocol.js";
+import { DefaultPubSubTopic } from "../../constants.js";
+import { groupByContentTopic } from "../../group_by.js";
+
+import {
+  FilterPushRpc,
+  FilterSubscribeResponse,
+  FilterSubscribeRpc,
+} from "./filter_rpc.js";
+
+const log = debug("waku:filter:v2");
+
+export type UnsubscribeFunction = () => Promise<void>;
+export type RequestID = string;
+
+type Subscription<T extends IDecodedMessage> = {
+  decoders: IDecoder<T>[];
+  callback: Callback<T>;
+  pubSubTopic: string;
+};
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+/**
+ * Implements client side of the [Waku v2 Filter protocol](https://rfc.vac.dev/spec/12/).
+ *
+ * Note this currently only works in NodeJS when the Waku node is listening on a port, see:
+ * - https://github.com/status-im/go-waku/issues/245
+ * - https://github.com/status-im/nwaku/issues/948
+ */
+class FilterV2 extends BaseProtocol implements IFilter {
```

Comment:
To me it seems `FilterV1` and `FilterV2` have a lot in common and we should find a way to decouple logic of js class `Filter` and underlying `rpc` so that later it's possible to do something like `new Filter(RpcV1)` or `new Filter(RpcV2);`.

Reply:
Filter v2 does give you more functionalities so we don't want to do an exercise or trying to simulate v2 functionalities on v1 either.

Reply:
How long we will keep `FilterV1`, @fryorcraken ?

Reply:
Until we are confident v2 is more stable than v1. ie, let's update our examples. Hopefully Status Mobile is also going to dogfood v2. Once done we cna start to deprecate v1. @jm-clius may have a plan already.